src/gitlab.js

```diff
@@ -115,25 +126,35 @@ export default class VerdaccioGitLab implements IPluginAuth {
       // In legacy mode, the groups are:
       // - for access, depending on the package settings in verdaccio
       // - for publish, the logged in user id and all the groups they can reach as fixed `$auth.gitlab.publish` = `$owner`
+      const gitlabAccessQueryParams = this.config.legacy_mode ? { owned: true } : { min_access_level: accessLevelId };
       const gitlabPublishQueryParams = this.config.legacy_mode ? { owned: true } : { min_access_level: publishLevelId };
       this.logger.trace('[gitlab] querying gitlab user groups with params:', gitlabPublishQueryParams);
 
-      const groupsPromise = GitlabAPI.Groups.all(gitlabPublishQueryParams).then(groups => {
+      const accessGroupsPromise = GitlabAPI.Groups.all(gitlabAccessQueryParams).then(groups => {
+        return groups.filter(group => group.path === group.full_path).map(group => group.path);
+      });
+      const publishGroupsPromise = GitlabAPI.Groups.all(gitlabPublishQueryParams).then(groups => {
         return groups.filter(group => group.path === group.full_path).map(group => group.path);
       });
 
-      const projectsPromise = GitlabAPI.Projects.all(gitlabPublishQueryParams).then(projects => {
+      const accessProjectsPromise = GitlabAPI.Projects.all(gitlabAccessQueryParams).then(projects => {
+        return projects.map(project => project.path_with_namespace);
+      });
+      const publishProjectsPromise = GitlabAPI.Projects.all(gitlabPublishQueryParams).then(projects => {
         return projects.map(project => project.path_with_namespace);
       });
 
-      Promise.all([groupsPromise, projectsPromise]).then(([groups, projectGroups]) => {
-        const realGroups = [user, ...groups, ...projectGroups];
-        this._setCachedUserGroups(user, password, { publish: realGroups });
+      Promise.all([accessGroupsPromise, publishGroupsPromise, accessProjectsPromise, publishProjectsPromise]).then(([accessGroups, publishGroups, accessProjectGroups, publishProjectGroups]) => {
+        const realAccessGroups = [user, ...accessGroups, ...accessProjectGroups];
+        const realPublishGroups = [user, ...publishGroups, ...publishProjectGroups];
+        this._setCachedUserGroups(user, {access: realAccessGroups, publish: realPublishGroups});
 
         this.logger.info(`[gitlab] user: ${user} successfully authenticated`);
-        this.logger.debug(`[gitlab] user: ${user}, with groups:`, realGroups);
+        this.logger.debug(`[gitlab] user: ${user}, with groups:`, realPublishGroups);
 
-        return cb(null, realGroups);
+        // The actual implementation of Verdaccio's plugin does not allow us to differentiate access and publish at this point.
+        // We'll use the cache to bypass that for now. For that reason, it does not matter what groups we return here.
+        return cb(null, [user]);
```

Comment:
I only return the username here, as I will be reading from the cache directly in the allow_publish and allow_access functions below.